### ISSUE TITLE
iperender segfault in cairo

### ISSUE THREAD

**Issue by lorenzhs** [otfried/ipe#342]
Hi Otfried,

I'm trying to iperender a presentation (ipe 7.2.20) and it segfaults. Traceback from a debug build (obtained by running `sed -i 's,-O2,,' src/common.mak`)

```gdb
(gdb) r
Starting program: /path/to/ipe-7.2.20/build/bin/iperender -pdf presentation.ipe export.pdf
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[Detaching after vfork from child process 100979]
Pdflatex was run sucessfully.

Program received signal SIGSEGV, Segmentation fault.
_cairo_image_compute_transparency (image=<optimized out>) at ../../../../src/cairo-image-surface.c:1199
1199	../../../../src/cairo-image-surface.c: No such file or directory.
(gdb) bt full
#0  _cairo_image_compute_transparency (image=<optimized out>) at ../../../../src/cairo-image-surface.c:1199
        a = <optimized out>
        pixel = 0x7ffff6336010
        x = 0
        y = 0
        transparency = CAIRO_IMAGE_IS_OPAQUE
        x = <optimized out>
        y = <optimized out>
        transparency = <optimized out>
        alpha = <optimized out>
        pixel = <optimized out>
        a = <optimized out>
#1  _cairo_image_compute_transparency (image=<optimized out>) at ../../../../src/cairo-image-surface.c:1158
        x = <optimized out>
        y = <optimized out>
        transparency = <optimized out>
        alpha = <optimized out>
        pixel = <optimized out>
        a = <optimized out>
#2  0x00007ffff790859d in _cairo_recording_surface_merge_source_attributes (op=<optimized out>, source=source@entry=0x5555555725f8, surface=<optimized out>, surface=<optimized out>)
    at ../../../../src/cairo-recording-surface.c:1779
        img_surf = 0x5555556db0a0
        surf_pat = 0x5555555725f8
        surf = 0x5555556db0a0
        free_me = 0x5555556db0a0
#3  0x00007ffff7908ce0 in _cairo_recording_surface_replay_internal (surface=<optimized out>, surface_extents=surface_extents@entry=0x0, surface_transform=surface_transform@entry=0x0, 
    target=target@entry=0x555555575870, target_clip=target_clip@entry=0x0, surface_is_unbounded=surface_is_unbounded@entry=0, type=<optimized out>, region=<optimized out>)
    at ../../../../src/cairo-recording-surface.c:2007
        command = 0x5555555725c0
        wrapper = {target = 0x555555575870, transform = {xx = 1, yx = 0, xy = 0, yy = 1, x0 = 0, y0 = 0}, has_extents = 1, extents = {x = 0, y = 0, width = 802, height = 602}, clip = 0x0, needs_transform = 0}
        elements = <optimized out>
        replay_all = <optimized out>
        status = CAIRO_INT_STATUS_SUCCESS
        extents = {x = 0, y = 0, width = 802, height = 602}
        use_indices = 0
        r = <optimized out>
        i = 0
        num_elements = 66
        __PRETTY_FUNCTION__ = "_cairo_recording_surface_replay_internal"
#4  0x00007ffff790a04b in _cairo_recording_surface_replay_and_create_regions (surface=<optimized out>, surface_transform=surface_transform@entry=0x0, target=target@entry=0x555555575870, 
    surface_is_unbounded=surface_is_unbounded@entry=0) at ../../../../src/cairo-recording-surface.c:2197
No locals.
#5  0x00007ffff78eac2c in _paint_page (surface=surface@entry=0x555555848af0) at ../../../../src/cairo-paginated-surface.c:417
        analysis = 0x555555575870
        status = CAIRO_INT_STATUS_SUCCESS
        has_supported = <optimized out>
        has_page_fallback = <optimized out>
        has_finegrained_fallback = <optimized out>
        __PRETTY_FUNCTION__ = "_paint_page"
#6  0x00007ffff78eb233 in _cairo_paginated_surface_show_page (abstract_surface=0x555555848af0) at ../../../../src/cairo-paginated-surface.c:583
        status = <optimized out>
        surface = 0x555555848af0
#7  0x00007ffff791fcd2 in INT_cairo_surface_show_page (surface=0x555555848af0) at ../../../../src/cairo-surface.c:2504
        status = CAIRO_STATUS_SUCCESS
#8  0x00007ffff78d7260 in _cairo_gstate_show_page (gstate=0x555555848ca0) at ../../../../src/cairo-gstate.c:1400
No locals.
#9  0x00007ffff792c315 in cairo_show_page (cr=0x555555848c70) at ../../../../src/cairo.c:2501
        status = <optimized out>
#10 0x00007ffff7fbb8a8 in ipe::Thumbnail::saveRender (this=0x7fffffffdc90, fm=ipe::Thumbnail::EPDF, dst=0x7fffffffe24e "export.pdf", page=0x5555555f24e0, view=0, zoom=1) at ipethumbs.cpp:189
        bbox = {iMin = {static ZERO = {static ZERO = <same as static member of an already seen type>, x = 0, y = 0}, x = -32.600000000000001, y = -69.599999999999994}, iMax = {static ZERO = {
              static ZERO = <same as static member of an already seen type>, x = 0, y = 0}, x = 768.60000000000002, y = 531.60000000000002}}
        wid = 802
        ht = 602
        data = {iData = std::shared_ptr<std::vector<char, std::allocator<char> >> (empty) = {get() = 0x0}}
        surface = 0x555555848af0
        file = 0x555555571b60
        cc = 0x555555848c70
        offset = {static ZERO = {static ZERO = <same as static member of an already seen type>, x = 0, y = 0}, x = -33, y = 532}
        painter = {<ipe::Painter> = {_vptr.Painter = 0x7ffff7fc8ce8 <vtable for ipe::CairoPainter+16>, iState = std::__cxx11::list = {[0] = {iStroke = {iRed = {iValue = 0}, iGreen = {iValue = 0}, iBlue = {
                    iValue = 0}}, iFill = {iRed = {iValue = 1000}, iGreen = {iValue = 1000}, iBlue = {iValue = 1000}}, iPen = {iValue = 1200}, iDashStyle = {static theEmptyString = 0x5555555719b0, 
                  iImp = 0x555555849430}, iLineCap = ipe::EButtCap, iLineJoin = ipe::ERoundJoin, iFillRule = ipe::EEvenOddRule, iSymStroke = {iRed = {iValue = 0}, iGreen = {iValue = 0}, iBlue = {iValue = 0}}, 
                iSymFill = {iRed = {iValue = 1000}, iGreen = {iValue = 1000}, iBlue = {iValue = 1000}}, iSymPen = {iValue = 1000}, iOpacity = {iValue = 1000}, iStrokeOpacity = {iValue = 1000}, iTiling = {
                  iName = 2147483648}, iGradient = {iName = 2147483648}}}, iMatrix = std::__cxx11::list = {[0] = {a = {1, 0, 0, 1, 0, 0}}}, iCascade = 0x555555571aa0, iAttributeMap = 0x5555555f2670, 
            iInPath = 0}, iFonts = 0x5555556e09f0, iCairo = 0x555555848c70, iZoom = 1, iPretty = true, iDimmed = false, iAfterMoveTo = false, iType3Font = false, iArgs = std::vector of length 0, capacity 8, 
          iResourceStack = std::vector of length 0, capacity 2, iPdfState = std::vector of length 9, capacity 16 = {{iStrokeRgb = {0, 0, 0}, iFillRgb = {0, 0, 0}, iStrokeOpacity = 1, iFillOpacity = 1, 
              iFont = 0x0, iFontSize = 1.4309223979077798e-197, iTextRise = 0, iCharacterSpacing = 0, iWordSpacing = 0, iHorizontalScaling = 1, iLeading = 0, iFillPattern = {
                static theEmptyString = 0x5555555719b0, iImp = 0x5555555719b0}}, {iStrokeRgb = {0, 0, 0}, iFillRgb = {0, 0, 0}, iStrokeOpacity = 1, iFillOpacity = 1, iFont = 0x5555556e9070, 
              iFontSize = 13.198499999999999, iTextRise = 0, iCharacterSpacing = 0, iWordSpacing = 0, iHorizontalScaling = 1, iLeading = 0, iFillPattern = {static theEmptyString = 0x5555555719b0, 
                iImp = 0x5555555719b0}}, {iStrokeRgb = {0, 0, 0}, iFillRgb = {0, 0, 0}, iStrokeOpacity = 1, iFillOpacity = 1, iFont = 0x5555556e9070, iFontSize = 8.2490000000000006, iTextRise = 0, 
              iCharacterSpacing = 0, iWordSpacing = 0, iHorizontalScaling = 1, iLeading = 0, iFillPattern = {static theEmptyString = 0x5555555719b0, iImp = 0x5555555719b0}}, {iStrokeRgb = {0, 0, 0}, 
              iFillRgb = {0, 0, 0}, iStrokeOpacity = 1, iFillOpacity = 1, iFont = 0x5555556e9070, iFontSize = 8.2490000000000006, iTextRise = 0, iCharacterSpacing = 0, iWordSpacing = 0, iHorizontalScaling = 1, 
              iLeading = 0, iFillPattern = {static theEmptyString = 0x5555555719b0, iImp = 0x5555555719b0}}, {iStrokeRgb = {0, 0, 0}, iFillRgb = {0, 0, 0}, iStrokeOpacity = 1, iFillOpacity = 1, iFont = 0x0, 
              iFontSize = 1.4309223979077798e-197, iTextRise = 0, iCharacterSpacing = 0, iWordSpacing = 0, iHorizontalScaling = 1, iLeading = 0, iFillPattern = {static theEmptyString = 0x5555555719b0, 
                iImp = 0x5555555719b0}}, {iStrokeRgb = {0, 0, 0}, iFillRgb = {0, 0, 0}, iStrokeOpacity = 1, iFillOpacity = 1, iFont = 0x555555859d70, iFontSize = 4.5827999999999998, iTextRise = 0, 
              iCharacterSpacing = 0, iWordSpacing = 0, iHorizontalScaling = 1, iLeading = 0, iFillPattern = {static theEmptyString = 0x5555555719b0, iImp = 0x5555555719b0}}, {iStrokeRgb = {0, 0, 0}, 
              iFillRgb = {0, 0, 0}, iStrokeOpacity = 1, iFillOpacity = 1, iFont = 0x555555859d70, iFontSize = 3.6661999999999999, iTextRise = 0, iCharacterSpacing = 0, iWordSpacing = 0, iHorizontalScaling = 1, 
              iLeading = 0, iFillPattern = {static theEmptyString = 0x5555555719b0, iImp = 0x5555555719b0}}, {iStrokeRgb = {0, 0, 0}, iFillRgb = {0, 0, 0}, iStrokeOpacity = 1, iFillOpacity = 1, 
              iFont = 0x5555556e9070, iFontSize = 8.2490000000000006, iTextRise = 0, iCharacterSpacing = 0, iWordSpacing = 0, iHorizontalScaling = 1, iLeading = 0, iFillPattern = {
                static theEmptyString = 0x5555555719b0, iImp = 0x5555555719b0}}, {iStrokeRgb = {0, 0, 0}, iFillRgb = {0, 0, 0}, iStrokeOpacity = 1, iFillOpacity = 1, iFont = 0x5555556e9070, 
              iFontSize = 9.1655999999999995, iTextRise = 0, iCharacterSpacing = 0, iWordSpacing = 0, iHorizontalScaling = 1, iLeading = 0, iFillPattern = {static theEmptyString = 0x5555555719b0, 
                iImp = 0x5555555719b0}}}, iTextMatrix = {a = {1, 0, 0, 1, 152.11229760000006, 785.20799999999997}}, iTextLineMatrix = {a = {1, 0, 0, 1, 0, 785.20799999999997}}}
        layerMatrices = std::vector of length 6, capacity 6 = {{a = {1, 0, 0, 1, 0, 0}}, {a = {1, 0, 0, 1, 0, 0}}, {a = {1, 0, 0, 1, 0, 0}}, {a = {1, 0, 0, 1, 0, 0}}, {a = {1, 0, 0, 1, 0, 0}}, {a = {1, 0, 0, 
              1, 0, 0}}}
#11 0x00005555555574fd in renderPage (fm=ipe::Thumbnail::EPDF, src=0x7fffffffe23d "presentation.ipe", dst=0x7fffffffe24e "export.pdf", pageSpec=0x0, viewSpec=0x0, zoom=1, transparent=false, nocrop=false)
    at iperender.cpp:78
        doc = 0x555555572d60
        pageIdx = 0
        page = 0x5555555f24e0
        viewIdx = 0
        tn = {iDoc = 0x555555572d60, iTransparent = false, iNoCrop = false, iWidth = 0, iHeight = 0, iZoom = 0, iLayout = 0x5555555e5338, iFonts = std::unique_ptr<ipe::Fonts> = {get() = 0x5555556e09f0}}
#12 0x00005555555579f8 in main (argc=4, argv=0x7fffffffde88) at iperender.cpp:176
        fm = ipe::Thumbnail::EPDF
        page = 0x0
        view = 0x0
        dpi = 72
        transparent = false
        nocrop = false
        i = 2
        src = 0x7fffffffe23d "presentation.ipe"
        dst = 0x7fffffffe24e "export.pdf"
```

I can send you the ipe file privately if it helps.

**Comment by otfried** [comment]
Can you mail me the `presentation.ipe` file?

Is there a specific reason you use iperender to generate a PDF file, instead of `ipetoipe -pdf -export` ?

**Comment by lorenzhs** [comment]
Just sent you the email. I have no good reason for using `iperender`, `ipetoipe` is the command I actually meant to use, but I only realised that I had the wrong command after `iperender` had already crashed ;)

**Comment by otfried** [comment]
Aha, fuzz testing :grinning: 

I got the file, thanks - it turns out the problematic part is using `\includegraphics` with a jpg file (it works for png).   I'm not sure yet if this is a Cairo bug or if Ipe is doing something wrong.

**Comment by lorenzhs** [comment]
Accidental fuzzing this time ;) Thanks for looking into it!

**Comment by otfried** [comment]
Fixed in 7.2.21.
